homeassistant/components/recorder/statistics.py

```diff
@@ -1134,38 +1141,41 @@ def reduce_week_ts_factory() -> (
     ]
 ):
     """Return functions to match same week and week start end."""
+    _boundries: tuple[float, float] = (0, 0)
+
     # We have to recreate _local_from_timestamp in the closure in case the timezone changes
     _local_from_timestamp = partial(
         datetime.fromtimestamp, tz=dt_util.DEFAULT_TIME_ZONE
     )
-    # We create _as_local_cached in the closure in case the timezone changes
-    _as_local_cached = lru_cache(maxsize=6)(_local_from_timestamp)
-
-    def _as_local_isocalendar(
-        time: float,
-    ) -> tuple:  # Need python3.11 for isocalendar typing
-        """Return the local isocalendar of a datetime."""
-        return _local_from_timestamp(time).isocalendar()
-
-    _as_local_isocalendar_cached = lru_cache(maxsize=6)(_as_local_isocalendar)
 
     def _same_week_ts(time1: float, time2: float) -> bool:
         """Return True if time1 and time2 are in the same year and week."""
-        date1 = _as_local_isocalendar_cached(time1)
-        date2 = _as_local_isocalendar_cached(time2)
-        return (date1.year, date1.week) == (date2.year, date2.week)  # type: ignore[attr-defined]
+        nonlocal _boundries
+        if not _boundries[0] <= time1 < _boundries[1]:
+            _boundries = _week_start_end_ts_cached(time1)
+        return _boundries[0] <= time2 < _boundries[1]
 
     def _week_start_end_ts(time: float) -> tuple[float, float]:
         """Return the start and end of the period (week) time is within."""
-        time_local = _as_local_cached(time)
+        nonlocal _boundries
+        time_local = _local_from_timestamp(time)
         start_local = time_local.replace(
             hour=0, minute=0, second=0, microsecond=0
         ) - timedelta(days=time_local.weekday())
-        start = dt_util.as_utc(start_local)
-        end = dt_util.as_utc(start_local + timedelta(days=7))
-        return (start.timestamp(), end.timestamp())
+        # We need to add 4 hours to the end to make sure we are in the next week
+        # before 0ing out the the time in case there is a DST change
```

Comment:
Adding 4 hours may be overkill, we may just need to do days. Need to check how this is handled over a DST boundry

Reply:
The four hours isn't needed
```
>>> datetime.datetime(year=2022,month=3,day=13,hour=0,minute=0,second=0,tzinfo=zoneinfo.ZoneInfo(key='america/chicago')) + datetime.timedelta(days=7)
datetime.datetime(2022, 3, 20, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='america/chicago'))
>>> datetime.datetime(year=2022,month=11,day=6,hour=0,minute=0,second=0,tzinfo=zoneinfo.ZoneInfo(key='america/chicago')) + datetime.timedelta(days=7)
datetime.datetime(2022, 11, 13, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='america/chicago'))
```